Deprecate _suggest endpoint in favour of _search

areek opened the pull request:
In 5.0, `_search` endpoint has been optimized for suggest
only search requests. Users should move away from
using the `_suggest` endpoint, as it may be deprecated and
removed in the future.

clintongormley commented:
I don't think we should suggest not using `_suggest` because it might be deprecated.  We should either deprecate it or not.  I'd be good with deprecating.

Mpdreamz commented:
+1 for deprecation, the dedicated endpoint returns only a marginally smaller response. 
If `_search` and `_suggest` now have same performance characteristics when only doing suggestions  `_suggest` now feels superfluous for the same reason we have no dedicated `/_aggs` endpoint :+1:

nik9000 commented on docs/reference/search/suggesters.asciidoc:
```diff
@@ -85,27 +91,35 @@ in the suggest text and if found an arbitrary number of options.
-  "my-suggest-2": ...
+  "hits": ...
+  "took": ...
+  "timed_out": ...
```
I'd just say `"timed_out": false,` and `"took": <whatever number it take on the first run on your laptop>,`. I like `...` for when you are eliding large structures but for short stuff like this I'd just let it be an example. You still need the `s///` clause for `took` but I think it makes the docs easier to read if the reader doesn't have to guess what expected values for `took` are.

You could also do something like:
```
  ...,
  "suggest": {
  }
```
To elide the whole search response. I don't think that is a particularly good idea here because it hides that this is a search response. But it is a valid strategy. You could do it with `// TESTRESPONSE [s/...,/"hits": "$body.hits", "took": "$body.took", "timed_out": "$body.timed_out"/]`.

areek commented:
>  I'd be good with deprecating.

++, I opened https://github.com/elastic/elasticsearch/pull/20435 deprecating _suggest endpoint.

areek commented:
@nik9000 Can you take look? In this PR:
 - updated all docs and rest-tests to use `_search` instead of `_suggest` endpoint
 - added deprecation logging to `_suggest` endpoint
 - added a rest test to ensure deprecation warning is set when `_suggest` is used

After this gets in, I will backport the change to 5.x and then remove `_suggest` endpoint entirely in master as planned in https://github.com/elastic/elasticsearch/pull/20435#issuecomment-267128683

nik9000 commented on rest-api-spec/src/main/resources/rest-api-spec/api/suggest.json:
```diff
@@ -1,7 +1,7 @@
   "suggest": {
     "documentation": "http://www.elastic.co/guide/en/elasticsearch/reference/master/search-suggesters.html",
-    "methods": ["POST", "GET"],
+    "methods": ["POST"],
```
It looks like GET is still supported. Maybe revert all changes to this file?

areek replied:
I removed the `GET` here to make sure we can test the deprecation warning [here](https://github.com/elastic/elasticsearch/pull/20305/files#diff-30d0b179e5896b66db3f509297d99cfdR33). Unfortunately, afaik the `warnings` message has to be an exact match.

nik9000 replied:
Ah! That makes sense. I believe it does have to match exactly.

nik9000 commented on rest-api-spec/src/main/resources/rest-api-spec/test/suggest/20_completion.yaml:
```diff
@@ -50,15 +50,16 @@ setup:
       indices.refresh: {}
 
   - do:
-      suggest:
```
Maybe it makes sense to move these to `test/search/110_completion.yaml`, etc.

areek replied:
thanks for the suggestion, on it

areek commented:
backported to [5.x](https://github.com/elastic/elasticsearch/commit/75489e2efaa04dfc22557a8aebfa4744d2a6b9fa)
